### ISSUE TITLE
PyPy support

### ISSUE THREAD

**Issue by Kobzol** [dask/distributed#231]
Is there any way to use `distributed` (specifically the workers, scheduler and executor is fine in Python) with PyPy?
I know that `sys.getsizeof` is not implemented in PyPy, so that's a problem, would it be possible to circumvent it somehow, so that the workers would not use it?

**Comment by mrocklin** [comment]
I was looking at this recently just for the scheduler and was surprised that things worked well.

I haven't looked at using PyPy workers/executor but in principle this should be easy enough to support.  The `sizeof` issue should be easy to hack around.  I expect that there will be other smaller issues though along the way.

Two questions:

1.  Is there a standard way to determine the size of an object in PyPy that we should use instead of `sys.getsizeof` or does PyPy not support this at all?  (not really a problem either way)
2.  I'm curious what your objective is with PyPy workers.  What sorts of computations are you doing?

**Comment by Kobzol** [comment]
I did a quick search and didn't find an alternative for `sizeof` in PyPy.
I tried monkey-patching distributed/sizeof.py with `sys.getsizeof = lambda x: 100` and I managed to get a computation done on a PyPy worker, but sometimes it seems that it gets stuck and no work is done (executor waits for a result and it doesn't come).
I want to do some number crunching (it's very basic stuff, iterating lists, dictionaries, doing maps, reduces etc.) and with PyPy I get about 4x speedup in 1 process vs. Python.
So I thought that it would be interesting to try running the workers in PyPy.

**Comment by mrocklin** [comment]
Yeah, if there is no alternative for `sys.`getsizeof` in PyPy then we would probably just implement something like your `lambda x: 100` solution, but instead directly in the `sizeof` functions rather than through monkeypatching.

My guess is that the other problem you're running into is unrelated.  There are probably a few small problems like this to solve.  I would need more information to help diagnose what this second issue is.  I would normally track these things down by running the test suite which is nice because it puts the executor, scheuler, and workers all in the same thread for easy debugging.

PyPy workers aren't very high on my priority list (most of my work is numeric) so I probably won't take the lead on this issue in the near future.  However I'd be happy to support you if you wanted to look more deeply into this.  I suspect that the next step is to run the test suite in PyPy.  In CPython I normally just run tests with `py.test distributed --verbose`.

**Comment by spirali** [comment]
> I'm curious what your objective is with PyPy workers. What sorts of computations are you doing?

Hello, I would like to answer the question as Kobzol's colleague.
We are working on a framework for questions in the area of automata theory;
one of our main goal is to generate some interesting instances for open problems in equivalence testing.

Our code is non-numeric and there is no single hotspot that would be easy to rewrite
into C/C++ and the whole thing is quite prototypical, so we would like to stick with Python.
These are the reasons why we are interested in Pypy, so we probably start some work in Pypy+Distributed direction.

**Comment by mrocklin** [comment]
Here are a few simple things to improve testing.  https://github.com/dask/distributed/pull/232

I ran into issues with the test `distributed/tests/test_executor.py::test_cancel` that surprised me.

**Comment by Kobzol** [comment]
I tried to run the test suite with PyPy and the newest master version.
There were some errors with bad file descriptors and missing S3FileSystem attributes, but the same errors appeared in Python tests, so that's probably not PyPy-related.

The only other errors that appeared were assertions about integers that received a long
(test_worker.py:75 → `assert isinstance(140407430174464L, int)`), sys.getsizeof usage in test_base and test_containers and the error you mentiond in test_cancel (but it only appears sometimes).

It appears that it is actually working quite nicely with PyPy, I'll try to use PyPy workers for some computations and see if there are any problems.

**Comment by mrocklin** [comment]
The s3fs and integer problems were resolved in #232 linked to above

**Comment by Kobzol** [comment]
Sorry, that was an error on my part, I pulled the commits but I reused the old test suite, so the integer fix wasn't there.
I'm still getting this error though:
`AttributeError: 'S3FileSystem' object has no attribute 'get_delegated_s3pars'` (both with Python 2.7.6 and PyPy 5.0.1).

**Comment by mrocklin** [comment]
You'll need to update or remove `s3fs`

**Comment by Kobzol** [comment]
Thanks, I removed it for the tests.
When I tried PyPy worker + Python sched/exec and PyPy worker/sched/exec, both gave me this error:
```
distributed.core - INFO - Failed to deserialize �cqit.base.transform
MapTransformation
q)�q}q(Ufnqccloudpickle.cloudpickle
_fill_function
q(ccloudpickle.cloudpickle
_make_skel_func
qccloudpickle.cloudpickle
_builtin_type
W|SqeTypeq�Rq   (KKKJCU%xtd�D]}||7}q
NM'�q
     Uxrangeq
UxqU␋─�qUtest.pyqU      transformqK
─))├─R─]─£─�─R─£─N£─RU⎻▒⎼␊┼├─␌─␋├.␉▒⎽␊.├⎼▒┼⎽°⎺⎼└
```
the rest was some weirdly-encoded traceback nonsense that totally messed up my terminal.
It turned out that I ran dworker and didn't set its PYTHONPATH to the source folders of my package, so it couldn't load the required packages for deserializing.
After setting the PYTHONPATH the computation went smoothly (and about 4x faster than in Python :-) ).

It seems that PyPy works well with your fixes, I'll be doing further tests and if I find something that doesn't work, I may report it here.
